File: src/divideConquer.js
/**
 * The divide step of the algorithm: recursively split a sorted site list into
 * singletons, building base bisectors for pairs and walking merge lines to
 * combine sibling halves.
 */

import {linkBisectorToSites, clearOutOrphans} from './bisector.js';
import {walkMergeLine, determineStartingBisector} from './mergeLine.js';

/**
 * Recursively split and merge sets of points.
 *
 * @param {Array<Site>} splitArray
 * @param {Metric} metric
 * @returns {Array<Site>}
 */
export function recursiveSplit(splitArray, metric){

    if(splitArray.length > 2){
        let splitPoint = (splitArray.length - splitArray.length % 2) / 2

        let L = recursiveSplit(splitArray.slice(0, splitPoint), metric);
        let R = recursiveSplit(splitArray.slice(splitPoint), metric);

        // Order the right-hand sites by distance from the merge start point.
        R.sort((a,b) => metric.distance(L[L.length - 1].site, a.site) - metric.distance(L[L.length - 1].site, b.site));

        let startingInfo = determineStartingBisector(L[L.length - 1], R[0], metric);

        let initialBisector = startingInfo.startingBisector;
        let initialR = startingInfo.nearestNeighbor;
        let initialL = startingInfo.w;

        let upStrokeArray = walkMergeLine(initialR, initialL, initialBisector, [metric.width, metric.height], true, metric);
        let downStrokeArray = walkMergeLine(initialR, initialL, initialBisector, [0, 0], false, metric);

        let mergeArray = [initialBisector, ...upStrokeArray, ...downStrokeArray];

        mergeArray.forEach(bisector => {
            bisector.mergeLine = splitArray.length;
            bisector.sites[0].bisectors = clearOutOrphans(bisector.sites[0], bisector.sites[1], metric);
            bisector.sites[1].bisectors = clearOutOrphans(bisector.sites[1], bisector.sites[0], metric);
            linkBisectorToSites(bisector);
        });

        return [...L, ...R];
    }
    else if(splitArray.length === 2){
        let bisector = metric.bisector(...splitArray);
        linkBisectorToSites(bisector);
        return splitArray;
    }
    else{
        return splitArray;
    }
}

File: py_ai/divide_conquer.py
"""
The divide step of the algorithm: recursively split a sorted site list into
singletons, building base bisectors for pairs and walking merge lines to
combine sibling halves.
"""

from bisector import link_bisector_to_sites, clear_out_orphans
from merge_line import walk_merge_line, determine_starting_bisector


def recursive_split(split_array, metric):
    """
    Recursively split and merge sets of points

    :param split_array: list of Site objects
    :param metric: Metric strategy object
    :returns: list of Site objects
    """
    # If more than two points, split recursively
    if len(split_array) > 2:
        split_point = (len(split_array) - len(split_array) % 2) // 2

        # Merge the child diagrams
        left = recursive_split(split_array[:split_point], metric)
        right = recursive_split(split_array[split_point:], metric)

        # The current working sites
        right_sorted = sorted(
            right, key=lambda s: metric["distance"](left[-1]["site"], s["site"])
        )

        starting_info = determine_starting_bisector(left[-1], right_sorted[0], metric)

        initial_bisector = starting_info["starting_bisector"]
        initial_r = starting_info["nearest_neighbor"]
        initial_l = starting_info["w"]

        up_stroke_array = walk_merge_line(
            initial_r,
            initial_l,
            initial_bisector,
            [metric["width"], metric["height"]],
            True,
            metric,
        )
        down_stroke_array = walk_merge_line(
            initial_r,
            initial_l,
            initial_bisector,
            [0, 0],
            False,
            metric,
        )

        # Combine all merge arrays
        merge_array = [initial_bisector] + up_stroke_array + down_stroke_array

        for bisector in merge_array:
            bisector["merge_line"] = len(split_array)
            bisector["sites"][0]["bisectors"] = clear_out_orphans(
                bisector["sites"][0], bisector["sites"][1], metric
            )
            bisector["sites"][1]["bisectors"] = clear_out_orphans(
                bisector["sites"][1], bisector["sites"][0], metric
            )
            link_bisector_to_sites(bisector)

        return left + right

    # Otherwise, determine the vertices if it has two sites
    elif len(split_array) == 2:
        bisector = metric["bisector"](split_array[0], split_array[1])
        link_bisector_to_sites(bisector)
        return split_array

    # If it has just one, just return it
    else:
        return split_array
